File: algorithms/0053.maximum_subarray/maximum_subarray.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut sum_list = vec![0; nums.len()];
        let mut prev = 0;
        for i in 0..nums.len() {
            sum_list[i] = prev + nums[i];
            prev = sum_list[i];
            if sum_list[i] < 0 {
                prev = 0;
            }
        }
        *sum_list.iter().max().unwrap_or(&0)
    }
}

File: algorithms/0053.maximum_subarray/maximum_subarray.cpp
class Solution {
public:
    int maxSubArray(const vector<int>& nums) {
        int maximum = nums[0];
        int sum = 0;
        for (const auto &n: nums) {
           sum += n;
           maximum = max(sum, maximum);
           if (sum < 0) {
               sum = 0;
           }
        }
        return maximum;
    }
};